### ISSUE TITLE
Investigate performance of action forms in 2.13

### ISSUE THREAD

**Issue by abbyad** [medic/medic-webapp#3837]
With 2.13.0-beta.10 it is taking ~5-12 seconds to load Action forms on standard.app (eg the Delivery form). It seems to take longer from the contact profile vs from the history tab. This could perhaps be related to `inputs` being populated, which contains a `db-object`.

We need to investigate this further to characterize the performance and determine if 2.13 can be released as such.

**Comment by garethbowen** [comment]
I've managed to eek out some small gains in the contact summary service but putting breakpoints in the code it looks like the bottleneck is Enketo rendering.

@abbyad Is there a 2.12 instance with the same forms we can use for comparison?

@alxndrsn Any ideas? Do you think it could be related to binding the contact summary into the form?

**Comment by alxndrsn** [comment]
<img width="461" alt="screen shot 2017-08-30 at 10 58 45" src="https://user-images.githubusercontent.com/191496/29864192-3cea2340-8d72-11e7-9070-2dd5ea4d30f1.png">

**Comment by garethbowen** [comment]
I've merged my caching improvements to master and we can backport if it'd help but I think there's something else going on that I haven't been able to detect yet...

**Comment by abbyad** [comment]
I will create and post links to forms today with minor variants that seem to have some effect on performance so that they can be profiled.

**Comment by abbyad** [comment]
Here are some [variants of the delivery form](https://github.com/medic/medic-webapp/files/1265197/issue-3837-variants-of-delivery-form.zip), which I then tested on `2.13.0-beta.10` on standard.app.

@alxndrsn offered to do some detailed profiling of the forms, but in the meantime here are some stopwatch results:

Form | Opened from History tab | Opened from Person's profile
--- | :---: | :---:
`issue-3837-0-no-change`              | 1.5s | 6s
`issue-3837-1-no-translations`        | 1s | 3.5s
`issue-3837-2-no-secondary-instance`  | 1s | 3.5s
`issue-3837-3-no-db-object`           | 1s | 3s

Opening the forms from the History tab does not run `contact-summary`, nor pass any info to the form, which is helpful for quick check before proper profiling is done tomorrow.

From this quick analysis it looks like removing unused `<translations>` has the biggest benefits, although not terribly practical for Standard projects. That said, it probably means this isn't anything new in 2.13 so may not be a blocker, but we can wait until the proper profiling before making that call.

**Comment by sglangevin** [comment]
Just a note that I did some quick testing on a Tecno Y4, v2.13.0-beta.10, container v0.3.6 and I don't see a noticeable difference between the contacts tab and history tab. I also don't think the performance has degraded significantly, but it is definitely slower to open forms than it was in the past. I also noticed that `contact-summary` seems to run when you open a form from the Tasks tab. Is that necessary to populate certain fields?

**Comment by alxndrsn** [comment]
# On remote server

The first five network requests when loading the [linked contact](https://standard.app.medicmobile.org/medic/_design/medic/_rewrite/#/contacts/B941731E-D532-F86E-B095-B788E9E67451) are impressively slow: 

<img width="919" alt="screen shot 2017-08-31 at 17 20 25" src="https://user-images.githubusercontent.com/191496/29931156-b740f034-8e70-11e7-988b-504a96a163a3.png">

**Comment by alxndrsn** [comment]
# Loading the actual form

<img width="951" alt="screen shot 2017-08-31 at 17 27 31" src="https://user-images.githubusercontent.com/191496/29931491-b38075ae-8e71-11e7-89ac-26e02c6ccf45.png">

N.B. this is with an all-access/fully-online user

## Queries
* why isn't the favicon cached?
* does the user object really (need to) get loaded twice simultaneously?

**Comment by alxndrsn** [comment]
# Profile of loading the delivery form on standard.app

<img width="261" alt="screen shot 2017-08-31 at 17 57 29" src="https://user-images.githubusercontent.com/191496/29932894-f18665b2-8e75-11e7-8115-267fd6861237.png">

=>

<img width="518" alt="screen shot 2017-08-31 at 17 58 55" src="https://user-images.githubusercontent.com/191496/29932934-13bc4278-8e76-11e7-86e5-ab209dae0989.png">

=>

## `.trigger()`

<img width="1016" alt="screen shot 2017-08-31 at 17 59 41" src="https://user-images.githubusercontent.com/191496/29932977-36a0d06a-8e76-11e7-8016-b1fd48b41ddf.png">
<img width="646" alt="screen shot 2017-08-31 at 18 00 15" src="https://user-images.githubusercontent.com/191496/29932998-4329a1b8-8e76-11e7-9ed4-0c0de30b3e01.png">

This looks like it's part of jQuery.

## `setVal()`

<img width="734" alt="screen shot 2017-08-31 at 18 00 43" src="https://user-images.githubusercontent.com/191496/29933013-56d771d6-8e76-11e7-853f-3c05b0db39cf.png">

This looks like it's part of enketo.

## `evaluate()`

<img width="1099" alt="screen shot 2017-08-31 at 18 02 35" src="https://user-images.githubusercontent.com/191496/29933078-9dbd4670-8e76-11e7-8d9b-a841693637e8.png">

This is part of our XPath evaluator.

**Comment by abbyad** [comment]
As requested by @alxndrsn, here is a version without the `output` fields in the labels: [issue-3837-4-no-outputs.zip](https://github.com/medic/medic-webapp/files/1268123/issue-3837-4-no-outputs.zip)

This can also be seen in action on standard.app

**Comment by alxndrsn** [comment]
Loading this form calls `jsEvaluate()` 2,500 times.  This implies 2.5k xpath expressions are evaluated...

**Comment by alxndrsn** [comment]
From a page refresh:

Form | `jsEvaluate()` count | Time to load on server
---|---
0 - No Change | 2452 | 15s
1 - No Translations | 856 | 8s
2 - No Secondary Instance | 856 | 6s
3 - No db-object | 170 | 7s
4 - No outputs | | 3s

Time to load was measured for Toronto Health Centre in Firefox after loading a few forms for the same user, including the form in question.  A single timing was taken.

## Observations

1. It seems a strange coincidence that **1** and **2** should have the same call count when the change made is so different, unless the only translated elements are in the secondary instance.
2. `db-object` is doing something pathological
3. A few things we do in our forms seem to severely affect enketo performance

**Comment by alxndrsn** [comment]
# Conclusion

## Removing the `jr:choice-name()` patch improves things _significantly_:

Form | `jsEvaluate()` count (with patch) | `jsEvaluate()` count (without patch)
---|---|---
0 - No Change | 2452 | 58
1 - No Translations | 856 | 58
2 - No Secondary Instance | 856 | 58
3 - No db-object | 170 | 54
4 - No outputs | 52 | 52

To improve performance I strongly recommend we find a replacement for our `jr:choice-name()` uses.

## References

* https://github.com/enketo/enketo-core/issues/412
* https://github.com/medic/medic-webapp/issues/3804
* https://github.com/medic/medic-webapp/issues/3281

**Comment by alxndrsn** [comment]
The final tests above were done on `master`.  I'm going to re-test this on the `2.13.x` branch, as the `jr:choice-name()` patch was slightly different in that release.  I expect similar results.

**Comment by alxndrsn** [comment]
I've manually removed the `jr:choice-name()` patch on `standard.app` and see form load times of around 5s for both `0 - No change` and `4 - No outputs`.

**Comment by abbyad** [comment]
After discussion with @alxndrsn we will attempt to get around the (over)use of `jr:choice-name()`, especially as used for `g_chw_sms` in the forms above. That calc sets the default value for a text field based on prior inputs, and the default value is a template that includes other values such as patient name, ID eg `Good news, ${chw_name}! ${patient_name} (${r_patient_id}) has delivered at the health facility.`

One strategy is to use `itext()` instead, as seen in this simplified example:
`<bind calculate="jr:itext( '/delivery/group_note/default_chw_sms/anc_only_home_sba_birth:label' )" nodeset="/delivery/group_note/g_chw_sms" type="string"/>`

However this gives the following error:
`Error loading form Error:  ["FormLogicError: Could not evaluate: jr:itext( '/delivery/group_note/default_chw_sms/anc_only_home_sba_birth:label' ),  message: Failed to execute 'evaluate' on 'Document': The string 'jr:itext(\"/delivery/group_note/default_chw_sms/anc_only_home_sba_birth:label\")' is not a valid XPath expression."]`

The form [issue-3837-0-5-no-choice-name.zip](https://github.com/medic/medic-webapp/files/1270935/issue-3837-0-5-no-choice-name.zip) can be seen in standard.app

**Comment by garethbowen** [comment]
@alxndrsn This is marked for "Code review" but I think it should be "Won't fix: Not an issue" because there's no code change. Does that make sense?

**Comment by alxndrsn** [comment]
I'd quite like someone to check my working 😄 

Also I don't think "not an issue" is a very helpful label when this clearly _is_ an issue.  Just not one whose solution can be tested or released in the standard way.

**Comment by alxndrsn** [comment]
@garethbowen did you check this before changing the status?

**Comment by garethbowen** [comment]
The logic looks fine to me. I put it in AT so the performance could be tested.

**Comment by alxndrsn** [comment]
> I put it in AT so the performance could be tested.
I haven't made any changes to code.

**Comment by abbyad** [comment]
The performance AT is blocked on https://github.com/medic/medic-webapp/issues/3870

**Comment by alxndrsn** [comment]
This ticket was to investigate the performance; I was requesting a review of my investigation.  Acceptance testing is to e.g. check the conclusions & recommendations I made.

**Comment by garethbowen** [comment]
As discussed in the planning meeting: the investigation is done, and performance improvements can be tested with #3870
